README.rst

```diff
@@ -87,6 +87,137 @@ There is built-in support for this option for Django >= 1.9. To use `sortkey`, s
 
+Using distkey
+---------------------------------
+
+There is built-in support for this option for Django >= 1.11. To use `distkey`, define an index on the model
+meta with the custom index type `django_redshift_backend.base.DistKey` with `fields` naming a single field::
+
+  class MyModel(models.Model):
+      ...
+
+      class Meta:
+          indexes = [DistKey(fields=['customer_id'])]
+
+Redshift doesn't have conventional indexes, and we don't generate SQL for them. We merely use
+`indexes` as a convenient place in the Meta to identify the `distkey`.
+
+You will likely encounter the following complication:
+
+Inlining Index Migrations
+~~~~~~~~~~~~~~~~~~~~~~~~~
+Django's `makemigrations` generates a migration file that first applies a `CreateModel` operation without the
+`indexes` option, and then adds the index in a separate `AddIndex` operation.
+
+However Redshift requires that the `distkey` be specified at table creation. As a result, you may need to
+manually edit your migration files to move the index creation into the initial `CreateModel`.
+
+That is, to go from::
+
+    operations = [
+        ...
+        migrations.CreateModel(
+            name='FactTable',
+            fields=[
+                ('distkeycol', models.CharField()),
+                ('measure1', models.IntegerField()),
+                ('measure2', models.IntegerField())
+                ...
+            ]
+        ),
+       ...
+       migrations.AddIndex(
+            model_name='facttable',
+            index=django_redshift_backend.base.DistKey(fields=['distkeycol'], name='...'),
+        ),
+    ]
+
+To::
+
+    operations = [
+        ...
+        migrations.CreateModel(
+            name='FactTable',
+            fields=[
+                ('distkeycol', models.CharField()),
+                ('measure1', models.IntegerField()),
+                ('measure2', models.IntegerField())
+                ...
+            ],
+            options={
+                'indexes': [django_redshift_backend.base.DistKey(fields=['distkeycol'], name='...')],
+            },
+        ),
+       ...
+    ]
+
+
+Inlining ForeignKey Migrations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+It is common to distribute fact tables on a foreign key column referencing the primary key of a dimension table.
+
+In this case you may also encounter the following added complication:
+
+Django's `makemigrations` generates a migration file that first applies a `CreateModel` operation without the
+`ForeignKey` column, and then adds the `ForeignKey` column in a separate `AddField` operation.  It does this to
+avoid attempts to create foreign key constraints against tables that haven't been created yet.
+
+However Redshift requires that the `distkey` be specified at table creation. As a result, you may need to
+manually edit your migration files to move the ForeignKey column into the initial `CreateModel`, while also
+ensuring that the referenced table appears *before* the referencing table in the file.
```

Comment:
Hum... I think editing migration files *before* migration is not a good experience. On the other hand, I feel it is useful to be able to automatically provided DISTKEY at migration phase.
@Surgo How do you think?

Reply:
The migration docs explicitly say that you're welcome to edit or even hand-craft migration files (and in some extreme cases you have to). `makemigrations` is just a convenience.

Reply:
This is not `ForeignKey` specific problems.

1. Get related fields like `ForeignKey` and `ManyToMany` from model
    https://github.com/django/django/blob/2.0.7/django/db/migrations/autodetector.py#L518-L534
1. Create table without related fields
    https://github.com/django/django/blob/2.0.7/django/db/migrations/autodetector.py#L557-L569
1. Add field for all related fields
    https://github.com/django/django/blob/2.0.7/django/db/migrations/autodetector.py#L575-L589

If we will not edit migration file, we have to override `generate_created_models`.

Reply:
Modifying the migration generation code seemed like it would potentially introduce instability, bugs etc. It might be nice to add in the future, but it's likely outside the scope of this change.

Reply:
No objection from me.